prune.go

```diff
@@ -242,47 +277,86 @@ type CleanupStats struct {
 	ShouldRetry bool
 }
 
+func startReportingThread(stop <-chan struct{}, progressCh chan ProgressMessage) {
+	tStart := time.Now()
+	next := time.After(15 * time.Second)
+	go func() {
+		txnsRemoved := 0
+		docsCleaned := 0
+		for {
+			select {
+			case <-stop:
+				return
+			case msg := <-progressCh:
+				txnsRemoved += msg.TxnsRemoved
+				docsCleaned += msg.DocsCleaned
+			case <-next:
+				txnRate := 0.0
+				since := time.Since(tStart).Seconds()
+				if since > 0 {
+					txnRate = float64(txnsRemoved) / since
+				}
+				logger.Debugf("pruning has removed %d txns (%.0ftxn/s) cleaning %d docs ",
+					txnsRemoved, txnRate, docsCleaned)
+				next = time.After(15 * time.Second)
+			}
+		}
+	}()
+}
+
 // CleanAndPrune runs the cleanup steps, and then follows up with pruning all
 // of the transactions that are no longer referenced.
 func CleanAndPrune(args CleanAndPruneArgs) (CleanupStats, error) {
+	tStart := time.Now()
 	var stats CleanupStats
 
 	if err := args.validate(); err != nil {
 		return stats, err
 	}
-
-	db := args.Txns.Database
-
-	if args.TxnsCount <= 0 {
-		txnsCount, err := args.Txns.Count()
-		if err != nil {
-			return stats, err
+	stop := make(chan struct{})
+	progressCh := make(chan ProgressMessage)
+	startReportingThread(stop, progressCh)
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var pstats PrunerStats
+	var anyErr error
+	prune := func(reversed bool) {
+		pruner := NewIncrementalPruner(IncrementalPruneArgs{
+			MaxTime:           args.MaxTime,
+			ProgressChannel:   progressCh,
+			ReverseOrder:      reversed,
+			TxnBatchSize:      args.TxnBatchSize,
+			TxnBatchSleepTime: args.TxnBatchSleepTime,
+		})
+		stats, err := pruner.Prune(args.Txns)
+		mu.Lock()
+		pstats = CombineStats(pstats, stats)
+		if anyErr == nil {
```

Comment:
Why only care about the first error hit by either goroutine?

Reply:
its mostly a case of 'what do we do with all the other errors'. probably we'd ideally want to at least log them. I went ahead and added that.